# ==========================================
# 1. IMPORT LIBRARIES
# ==========================================

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# ==========================================
# 2. LOAD & PREPARE DATA
# ==========================================

In [2]:
# The dataset has no header, and column 0 is the target.
print("Loading data...")
df = pd.read_csv('midterm-regresi-dataset.csv', header=None)

# Target is the first column, features are the rest
y = df.iloc[:, 0]
X = df.iloc[:, 1:]

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Scale the data (Crucial for deep learning and distance-based ML)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

Loading data...


# ==========================================
# 3. MACHINE LEARNING: RANDOM FOREST REGRESSOR
# ==========================================

In [3]:
print("Training Random Forest Regressor...")
rf_regressor = RandomForestRegressor(n_estimators=100, max_depth=15, random_state=42, n_jobs=-1)
rf_regressor.fit(X_train_scaled, y_train)

# Predictions
rf_preds = rf_regressor.predict(X_test_scaled)

# Evaluation
print("\n--- Random Forest Regression Evaluation ---")
print("MSE:", mean_squared_error(y_test, rf_preds))
print("RMSE:", np.sqrt(mean_squared_error(y_test, rf_preds)))
print("MAE:", mean_absolute_error(y_test, rf_preds))
print("R2 Score:", r2_score(y_test, rf_preds))

Training Random Forest Regressor...

--- Random Forest Regression Evaluation ---
MSE: 84.14088096409212
RMSE: 9.172833856780146
MAE: 6.445725475290161
R2 Score: 0.29302300718412566


# ==========================================
# 4. DEEP LEARNING REGRESSION MODEL
# ==========================================

In [4]:
print("\nTraining Deep Learning Regressor...")
dl_regressor = Sequential([
    Dense(128, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(1) # Linear activation for regression output
])

dl_regressor.compile(optimizer='adam', loss='mse', metrics=['mae'])

# Train
history = dl_regressor.fit(X_train_scaled, y_train, epochs=10, batch_size=128, validation_split=0.2, verbose=1)

# Predict and Evaluate
dl_preds = dl_regressor.predict(X_test_scaled).flatten()

print("\n--- Deep Learning Regression Evaluation ---")
print("MSE:", mean_squared_error(y_test, dl_preds))
print("RMSE:", np.sqrt(mean_squared_error(y_test, dl_preds)))
print("MAE:", mean_absolute_error(y_test, dl_preds))
print("R2 Score:", r2_score(y_test, dl_preds))


Training Deep Learning Regressor...


C:\Users\RAKHAP\AppData\Roaming\Python\Python313\site-packages\keras\src\layers\core\dense.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
2577/2577 ━━━━━━━━━━━━━━━━━━━━ 4s 1ms/step - loss: 228409.1406 - mae: 253.2668 - val_loss: 2647.4229 - val_mae: 30.6310
Epoch 2/10
2577/2577 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - loss: 848.9250 - mae: 17.9009 - val_loss: 536.5597 - val_mae: 15.3850
Epoch 3/10
2577/2577 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - loss: 243.8340 - mae: 10.6465 - val_loss: 268.7150 - val_mae: 10.4908
Epoch 4/10
2577/2577 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - loss: 173.6589 - mae: 9.4804 - val_loss: 177.4876 - val_mae: 9.4445
Epoch 5/10
2577/2577 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - loss: 204.3719 - mae: 9.8241 - val_loss: 130.9606 - val_mae: 8.0973
Epoch 6/10
2577/2577 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - loss: 172.5901 - mae: 9.1151 - val_loss: 301.7156 - val_mae: 13.8701
Epoch 7/10
2577/2577 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - loss: 153.5281 - mae: 8.9378 - val_loss: 121.4513 - val_mae: 8.3047
Epoch 8/10
2577/2577 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - loss: 157.2317 - mae: 8.9198 - val_loss: 139.3247 - val_mae: 8.3

### **Conclusion and Interpretation of Results**

Based on the evaluation metrics, the traditional Machine Learning model (Random Forest) significantly outperformed the Deep Learning model (Sequential Neural Network) for this specific regression task.

**1. Random Forest Regressor Performance:**
*   The model achieved a Mean Absolute Error (MAE) of ~6.45. This indicates that, on average, the model's prediction for a song's release year is off by approximately 6.5 years.
*   The R² score is ~0.293. While this implies the model only captures about 29.3% of the variance in the target variable, it demonstrates that the model is successfully learning underlying patterns from the audio features better than a simple baseline prediction.

**2. Deep Learning Regressor Performance:**
*   The Neural Network performed poorly, yielding a negative R² score of -0.428. In regression analysis, a negative R² score signifies that the model's predictions are worse than a horizontal line that simply predicts the average release year for every song.
*   The MAE (~9.13) and RMSE (~13.04) are considerably higher than those of the Random Forest model, confirming a higher rate of error in its predictions.

**Final Verdict:**
For this dataset and current baseline configuration, the **Random Forest Regressor is the superior model**. The Deep Learning model's failure to converge properly suggests that its current architecture or training parameters are not suited for this specific data structure. To improve the neural network's performance, future steps would require extensive hyperparameter tuning, such as adjusting the learning rate, modifying the number of hidden layers or neurons, applying regularization techniques, or training for more epochs.